keps/sig-network/2091-cluster-network-policy/README.md

```diff
@@ -0,0 +1,1161 @@
+| Rule identifiers         | Name per rule in string format. Unique within a CNP.                                                             | Not supported                                                                      |
+| Cluster external traffic | Not supported                                                                                                    | Supported via IPBlock                                                              |
+| Namespace selectors      | Supports advanced selection of Namespaces with the use of `namespaces`and label based `namespaceSelector` fields | Supports label based Namespace selection with the use of `namespaceSelector` field |
+
+### Notes/Constraints/Caveats
+
+It is important to note that the controller implementation for cluster-scoped
+policy APIs will not be provided as part of this KEP. Such controllers which
+realize the intent of these APIs will be provided by individual network policy
+providers, as is the case with the NetworkPolicy API.
+
+### Risks and Mitigations
+
+A potential risk of the ClusterNetworkPolicy resource is, when it's stacked on
+top of existing NetworkPolicies in the cluster, some existing allowed traffic
+patterns (which were regulated by those NetworkPolicies) may become blocked by
+ClusterNetworkPolicy `Deny` rules, while some isolated workloads may become
+accessible instead because of ClusterNetworkPolicy `Allow` rules.
+
+Developers could face some difficulties figuring out why the NetworkPolicies
+did not take effect, even if they know to look for ClusterNetworkPolicy rules
+that can potentially override these policies:
+To understand why traffic between a pair of Pods is allowed/denied, a list of
+NetworkPolicy resources in both Pods' Namespace used to be sufficient
+(considering no other CRDs in the cluster tries to alter traffic behavior).
+The same Pods, on the other hand, can appear as an AppliedTo, or an ingress/egress
+peer in any ClusterNetworkPolicy. This makes looking up policies that affect a
+particular Pod more challenging than when there's only NetworkPolicy resources.
+
+In case where a ClusterNetworkPolicy `Pass` rule,
+ClusterNetworkPolicy `Deny` rule, ClusterNetworkPolicy `Allow` rule,
+NetworkPolicy rule and default rule applies to an overlapping
+set of Pods, users will need to refer to the priority associated with the
+rule, to determine which rule would take effect. Figuring out how stacked policies
+affect traffic between workloads might not be very straightfoward.
+
+To mitigate this risk and improve UX, a tool which does a reverse look up and lists rules
+and policies that impact a given Pod and prints out relative precedence of those rules
+can be quite useful. The [cyclonus](https://github.com/mattfenwick/cyclonus)
```

Comment:
I think this needs to be done, but I don't think it should be the CNI, this process should stay in tree and not be implementation specific.

Reply:
Admin -> Needs to know what my CNP is going to break and has the ability to see that 

Developer -> Needs to know if my NPs are going to be overridden/ interrupted by any CNPS

Reply:
Need to add more detail to this section

Reply:
So... I guess even without special privileges, we can always answer the question "Is this connection allowed or denied according to _my_ NetworkPolicies?".

Like, if I'm an admin of Namespace A, and I want my pod "foo" to be able to connect to service "bar" in Namespace B, then I ask this tool, and it would say "according to the NPs in Namespace A, this connection is allowed, but it might be blocked by an NP in Namespace B, or a CNP". So then, if the connection _doesn't_ work, at least you know that it's not your fault.

So then you could go to a Namespace B admin and have them run the same check, and the tool would say "according to the NPs in Namespace B, this connection is allowed, but it might be blocked by an NP in Namespace A, or a CNP". And so then at that point, you know it must be blocked by a CNP.

(And a cluster-administrator could run the tool, and it would say "this connection is allowed by the NPs in Namespace A and the NPs in Namespace B but is blocked by CNP X".)